feat: broken simpprocs for the LLVMAndRiscV dialect

salinhkuhn opened the pull request:
This PR adds simprocs for the LLVMAndRiscV dialect so that simp_peephole can recognize and simplify specific patterns. However, the current simprocs are not yet sufficient. In particular, simplification fails in the case of cast instructions—we are unable to eliminate the framework cast, and instead encounter a "move not type correct" error. This issue may be related to the newly introduced LLVM parser.

Commit:
refactor: mv LLVMAndRiscV into separate folder

Commit:
Merge remote-tracking branch 'origin/main' into LLVMAndRiscv64Refinement

Commit:
feat: Peephole Refine

Commit:
fix: line length

Commit:
fix: unsupported op to generic error

Commit:
added comment

Commit:
test case

Commit:
merge

Commit:
added simproc debug test cases

Commit:
Merge remote-tracking branch 'origin/main' into simpprocs

Commit:
Merge remote-tracking branch 'origin/main' into simpprocs

Commit:
temp simpproc

Commit:
simpproc issues

Commit:
borken simpprocs

Commit:
Merge remote-tracking branch 'origin/main' into simpprocs

Commit:
fix: line length

Commit:
fix: identation

salinhkuhn commented:
In combination with the PR  #1247 the simpprocs successfully simplify the framework overhead. Thus they still require to be merged.

tobiasgrosser commented:
They are merged now. Can you click "update branch" and prepare this PR for a final review?

salinhkuhn committed:
Merge branch 'main' into simpprocs

Commit:
clean up simpprocs

Commit:
rmv lemmas

alexkeizer commented on SSA/Projects/LLVMRiscV/simpproc.lean:
```diff
@@ -0,0 +1,144 @@
+| .cons x xs => do
+  let tailExpr ← listExprToExprShallow type xs
+  mkAppOptM ``List.cons #[type, x, tailExpr]
+
```
This feels like it'd ought to exist in the standard library already. A loogle search for the signature shows [`Lean.Meta.mkListLit`](https://leanprover-community.github.io/mathlib4_docs/Lean/Meta/AppBuilder.html#Lean.Meta.mkListLit), whose [implementation](https://github.com/leanprover/lean4/blob/81b85203c90439fb7d7e0cec32130d62d8cb7a2e/src/Lean/Meta/AppBuilder.lean#L517-L524) looks like it does the right thing. Can you check whether replacing listExprToShallow with Lean.Meta.mkListLit (and dropping the former) also works?

salinhkuhn replied:
Thx, I managed to replace the function with the already given implementation.

alexkeizer commented on SSA/Projects/LLVMRiscV/simpproc.lean:
```diff
@@ -0,0 +1,144 @@
+simproc [simp_denote] llvmArgsFromHybrid_cons_eq (llvmArgsFromHybrid (_) ) := fun e => do
+-- TODO: (comment from Alex)This ought to be .cons rather than a wildcard -------------^^^
+--       Unfortunately, then it the simproc again stops being applied when we'd
+--       expect it to, despite liberal use of `no_index`.
```
```suggestion
-- TODO: This ought to be .cons rather than a wildcard -------------^^^
--       Unfortunately, then it the simproc again stops being applied when we'd
--       expect it to, despite liberal use of `no_index`.
```

Note I took care so that the `^^^` symbols point towards the "this" the comment referenced, please don't break that!

alexkeizer commented on SSA/Projects/LLVMRiscV/simpproc.lean:
```diff
@@ -0,0 +1,144 @@
+
+/- The following additional lemmas are needed to simplify the proof goals for the hybrid dialect.-/
+@[simp]
+theorem poisonOr_mk_some_eq_value (x : α) : { toOption := some x } = PoisonOr.value x := rfl
```
Can you move this to the Poison.lean file, and name it `mk_some` (in the `PoisonOr` namespace). While you're at it, please also add the `mk_none` analogous lemma!

salinhkuhn replied:
done

alexkeizer commented on SSA/Projects/LLVMRiscV/simpproc.lean:
```diff
@@ -0,0 +1,144 @@
+theorem poisonOr_mk_some_eq_value (x : α) : { toOption := some x } = PoisonOr.value x := rfl
+
+@[simp_denote]
+theorem liftM_eq_some (α : Type u) : @liftM Id Option _  α = some := by rfl
```
Do you still need this? I thought we got rid of the superfluous option that got introduced?

alexkeizer reviewed (changes requested):
Could you write a comment that (to the best of your understanding; it's fine if this is not quite perfect) explains why we need the simprocs?

Commit:
Merge remote-tracking branch 'origin/main' into simpprocs

salinhkuhn commented:
> Could you write a comment that (to the best of your understanding; it's fine if this is not quite perfect) explains why we need the simprocs?

I added a comment at the top of the file.
I implemented all your comments, therefore the PR would be ready for a second round of review @alexkeizer

Commit:
mv lemma to Poison file, add comment

alexkeizer reviewed (approved):
LGTM, although I don't really like that we have to resort to simprocs for this, and the current simprocs could likely be cleaned up, it's probably not worth to spend the effort that would take, so this hack shall live on.

(For posterity (i.e., git-blame purposes), I believe these simprocs were originally authored by @bollu)
